Provide an error message when trying to use wxGLCanvas on non-X11 dis…

swt2c opened the pull request:
…play

Currently, wxGLCanvas on wxGTK is only supported on X11 displays.  If a user
attempts to use wxGLCanvas on Wayland, for example, the application will core
dump.  This change adds an error message and a suggested workaround until
wxGLCanvas is supported on Wayland.

See https://trac.wxwidgets.org/ticket/17702

vadz commented on src/gtk/glcanvas.cpp:
```diff
@@ -350,4 +354,15 @@ void wxGLCanvas::GTKInitImplicitContext()
+    if ( !GDK_IS_X11_DISPLAY(gdk_display_get_default()) )
+#endif
+    {
+        wxString msg(wxT("wxGLCanvas is only supported on X11 currently.  You may be able to\nwork around this by setting environment variable GDK_BACKEND=x11 before starting\nyour program."));
```
Why use this temporary string instead of passing it directly to `wxLogError()`?

Also, this should normally be translated, i.e. use `_(...)` around it.

vadz commented on src/gtk/glcanvas.cpp:
```diff
@@ -350,4 +354,15 @@ void wxGLCanvas::GTKInitImplicitContext()
+#endif
+    {
+        wxString msg(wxT("wxGLCanvas is only supported on X11 currently.  You may be able to\nwork around this by setting environment variable GDK_BACKEND=x11 before starting\nyour program."));
+        wxLogFatalError(msg);
```
This is, of course, an improvement compared to simply crashing, but could we just return false from here and from `Create()` above and just return to the application after logging a (non-fatal) error? This would seem to be a bit more developer-friendly than just killing the program...

swt2c replied:
In that case, the constructor would just ignore the return value of `Create()` and it would appear as if the construction worked.  Then, I suspect the application would just crash at some later point when it tried to use the `wxGLCanvas.`

vadz replied:
This is not pretty, to put it mildly, but the canonical way to check for window creation success with wx API is to check for `win->GetHandle() != 0` after creating it. So if this still works here, it's fine.

Even if not, I think it's still better to at least allow using `Create()` without terminating the application...

vadz commented on include/wx/gtk/glcanvas.h:
```diff
@@ -111,6 +111,8 @@ class WXDLLIMPEXP_GL wxGLCanvas : public wxGLCanvasX11
     // implementation from now on
     void OnInternalIdle() wxOVERRIDE;
 
+    void EnsureAvailable();
```
Not sure if we want to expose this hack in the header, this could be a private function in `src/gtk/glcanvas.cpp` just as well AFAICS.

vadz reviewed (commented):
Thanks for trying to improve this but using `wxLogFatalError()` is not great IMHO, some programs could continue running even if they can't create their OpenGL windows.

Could this be improved further?

swt2c committed:
Provide an error message when trying to use wxGLCanvas on non-X11 display

Currently, wxGLCanvas on wxGTK is only supported on X11 displays.  If a user
attempts to use wxGLCanvas on Wayland, for example, the application will core
dump.  This change adds an error message and a suggested workaround until
wxGLCanvas is supported on Wayland.

See https://trac.wxwidgets.org/ticket/17702

swt2c commented:
Okay, I think I've addressed all of your comments.  I ended up having to use `wxSafeShowMessage()` because the samples would crash before `wxLogError()` would work.

vadz commented:
Thanks for the update(s)!

swt2c commented:
Can you cherry-pick to 3.0 as well?  Thanks!
